Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift

```diff
@@ -117,13 +152,47 @@ extension SymbolGraph.Symbol {
     public var isOverloadGroup: Bool {
         self.identifier.precise.hasSuffix(Self.overloadGroupIdentifierSuffix)
     }
+
+    /// Returns a sort comparator suitable for sorting symbols within an overload group.
+    ///
+    /// The sort order for overloads follows the following algorithm:
+    /// - If `orderByDeclaration` is false, symbols are sorted by their precise identifier.
+    /// - If `orderByDeclaration` is true, symbols are sorted by their ``DeclarationFragments``
+    ///   mixin, falling back to their precise identifier if their declarations are equal.
+    internal static func sortForOverloads(orderByDeclaration: Bool) -> ((Self, Self) -> Bool) {
+        return { isDeclarationBefore(
+            $0, $1, orderByDeclaration: orderByDeclaration,
+            getDeclaration: { $0.declarationFragments!.rendered },
```

Comment:
The reason i went for this version was so that i could have one single `sorted(by:)` call where it's needed and perform the branching here in the underlying implementation. If there are separate `sortForOverloadsByDeclaration` and `sortForOverloadsByIdentifier` calls, the callsites still look like this, where a variable is initialized within an if-branch rather than inline:

https://github.com/apple/swift-docc-symbolkit/blob/8a740b5197ece51e4b597f8a467546da66324581/Sources/SymbolKit/SymbolGraph/SymbolGraph%2BOverloads.swift#L54-L59

I can see the utility of not force-unwrapping here, but it acts like a precondition: If we are trying to sort by declaration, there _must_ be a declaration available to sort by. I don't want to fall back to sorting by identifier at this point in the stack because it can create inconsistent or confusing results in the resulting sort. I can create an actual `precondition(_:_:)` call so that it emits a real error message in case of misuse, but i believe that removing the force-unwrap entirely will cause more problems than it solves.

> I wonder also if there's a generally simpler way to express this logic, not having to pass in the two different closures with the boolean flag which is a bit awkward.

I can see about simplifying it with a protocol, but by having the actual comparison in a single place like this, i can guarantee that the sorting will behave the same between single symbols and unified symbols, which was the cause of some bugs as i was testing this code.

Reply:
Ok makes sense. The force-unwrap itself wouldn't worry me if the allSatisfy check was in the same function. But now it's sort of an undocumented requirement that callers might not know about.